drivers/usb/usbehci_new/roothub.c

```diff
@@ -0,0 +1,710 @@
+/*
+ * PROJECT:     ReactOS USB EHCI Miniport Driver
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     USBEHCI root hub functions
+ * COPYRIGHT:   Copyright 2017-2018 Vadim Galyant <vgal@rambler.ru>
+ */
+
+#include "usbehci.h"
+
+#define NDEBUG
+#include <debug.h>
+
+#define NDEBUG_EHCI_ROOT_HUB
+#include "dbg_ehci.h"
+
+MPSTATUS
+NTAPI
+EHCI_RH_ChirpRootPort(IN PVOID ehciExtension,
+                      IN USHORT Port)
+{
+    PEHCI_EXTENSION EhciExtension = ehciExtension;
+    PULONG PortStatusReg;
+    EHCI_PORT_STATUS_CONTROL PortSC;
+    ULONG PortBit;
+    ULONG ix;
+
+    DPRINT_RH("EHCI_RH_ChirpRootPort: Port - %x\n", Port);
+    ASSERT(Port != 0);
+
+    PortStatusReg = &EhciExtension->OperationalRegs->PortControl[Port - 1].AsULONG;
+    PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+    DPRINT_RH("EHCI_RH_ChirpRootPort: PortSC - %X\n", PortSC.AsULONG);
+
+    PortBit = 1 << (Port - 1);
+
+    if (PortBit & EhciExtension->ResetPortBits)
+    {
+        DPRINT_RH("EHCI_RH_ChirpRootPort: Skip port - %x\n", Port);
+        return MP_STATUS_SUCCESS;
+    }
+
+    if (PortSC.PortPower == 0)
+    {
+        DPRINT_RH("EHCI_RH_ChirpRootPort: Skip port - %x\n", Port);
+        return MP_STATUS_SUCCESS;
+    }
+
+    if (PortSC.CurrentConnectStatus == 0 ||
+        PortSC.PortEnabledDisabled == 1 ||
+        PortSC.PortOwner == EHCI_PORT_OWNER_COMPANION_CONTROLLER)
+    {
+        DPRINT_RH("EHCI_RH_ChirpRootPort: No port - %x\n", Port);
+        return MP_STATUS_SUCCESS;
+    }
+
+    if (PortSC.LineStatus == EHCI_LINE_STATUS_K_STATE_LOW_SPEED &&
+        PortSC.Suspend == 0 &&
+        PortSC.CurrentConnectStatus == 1)
+    {
+        /* Attached device is not a high-speed device.
+           Release ownership of the port to a selected HC.
+           Companion HC owns and controls the port. Section 4.2 */
+        PortSC.PortOwner = EHCI_PORT_OWNER_COMPANION_CONTROLLER;
+        WRITE_REGISTER_ULONG(PortStatusReg, PortSC.AsULONG);
+
+        DPRINT_RH("EHCI_RH_ChirpRootPort: Companion HC port - %x\n", Port);
+        return MP_STATUS_SUCCESS;
+    }
+
+    DPRINT("EHCI_RH_ChirpRootPort: EhciExtension - %p, Port - %x\n",
+           EhciExtension,
+           Port);
+
+    PortSC.PortEnabledDisabled = 0;
+    PortSC.PortReset = 1;
+    WRITE_REGISTER_ULONG(PortStatusReg, PortSC.AsULONG);
+
+    RegPacket.UsbPortWait(EhciExtension, 10);
+
+    do
+    {
+        PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+
+        PortSC.ConnectStatusChange = 0;
+        PortSC.PortEnableDisableChange = 0;
+        PortSC.OverCurrentChange = 0;
+        PortSC.PortReset = 0;
+
+        WRITE_REGISTER_ULONG(PortStatusReg, PortSC.AsULONG);
+
+        for (ix = 0; ix <= 500; ix += 20)
+        {
+             KeStallExecutionProcessor(20);
```

Comment:
Is it a good idea to use `KeStallExecutionProcessor`, since it totally stops the processor for specified amount of time?

And why not to use `KeDelayExecutionThread` instead?

Reply:
KeDelayExecutionThread incurs scheduling overhead by giving up the thread's current time slice. The next time it gets scheduled will be something like 10-20ms later -- which is order of magnitudes longer than the intended sleep of 20 microseconds here. See https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-kestallexecutionprocessor